### PR TITLE
Lesson3

### PR DESCRIPTION
**Pull request by VeronikaVasilchikova**
Date: 2020/03/09

The following fetures were implemented here:
- Create collections for Contacts, Statuses and Countries
- Sync UI elements with the corresponding collections
- Connect all create/update/delete operations to backend

Unfortunately, this server didn't want to work at my computer but I tried to do this task.
I'll be gald to get your feedback (whether I think in the right direction?)

### CODE REVIEW

File: sources/views/contacts.js [hunk 1]
```diff
@@ -27,8 +27,9 @@ export default class ContactsView extends JetView {
 									select: true,
 									onClick: {
 										"wxi-close": function(e, id) {
-											contacts.remove(id);
-											return false;
+											contacts.waitData.then(() => {
```
Comment (Zwillinge) [inline sources/views/contacts.js]: no need to wrap in waitData - users cannot press they close icon until data is loaded

File: sources/views/contacts.js [hunk 2]
```diff
@@ -46,8 +47,10 @@ export default class ContactsView extends JetView {
 		};
 	}
 	doAddClick(){
-		const id = contacts.add({Name: "", Email: "", Status: "", Country: ""}, 0);
-		this.listOfUsers.select(id);
+		contacts.waitData.then(() => {
```
Comment (Zwillinge) [inline sources/views/contacts.js]: waitData promise is used to wait for  loaded data, not for save operation
~~~
contacts.waitSave(()=>{
    contacts.add({...})
}).then((res)=>{
    list.select(res.id); //you need to wait for server response to get the correct id of a new item
})
~~~

File: sources/views/form.js [hunk 3]
```diff
@@ -51,12 +51,14 @@ export default class ContactsView extends JetView {
 							label:"Save",
 							type:"form",
 							click:() => {
-								const values = this.getRoot().getValues();
-								if(values.id && (values.Name != "" || values.Email != "")) {
-									contacts.updateItem(values.id, values);
-								} else {
-									webix.message("Select other contact to start editing");
-								}
+								contacts.waitData.then(() => {
```
Comment (Zwillinge) [inline sources/views/form.js]: no need to wrap this operation in waitData  - the form will be filled and this operation will become possible only when data is loaded. you need to wait for data only if you want to perform smth immediately after loading

File: sources/views/contacts.js (hunk 2 above)
Comment (VeronikaVasilchikova) [reply sources/views/contacts.js]: Thanks for you advice! I fixed id

**Comment by Zwillinge** [comment]
and please double-check the following thing: 

when you work with countries/statuses in GridView module, you need to add and delete to a collection, not datatable. you can easily see now that combo boxes in a form  should react  to changes in these collections as they are also synced

File: sources/views/form.js [hunk 4]
```diff
@@ -26,7 +26,7 @@ export default class ContactsView extends JetView {
 					placeholder:"Options",
 					options: {
 						body: {
-							data: countries,
+							collection: countries,
```
Comment (Zwillinge) [inline sources/views/form.js]: does it really work? please check it on real app, not just fix the code. 

to sync combo options with a collection, you need to use the same `data` property in its suggest body

Comment (VeronikaVasilchikova) [reply sources/views/form.js]: You were right. It doesn't work when I write "collection". When I returned to previous variant (collection=>data) it started to work.

**Pull request merged**